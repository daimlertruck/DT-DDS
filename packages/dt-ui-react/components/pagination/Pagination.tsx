import React, { useEffect, useMemo, useState } from 'react';

import { ArrowLeftIcon, ArrowRightIcon } from '../../core';
import { BaseProps } from '../../types';
import { debounce } from '../../utils';
import { Typography } from '../typography';

import {
  PaginationStyled,
  InputStyled,
  LastPageStyled,
  ArrowStyled,
} from './Pagination.styled';

export interface ContentProps {
  totalPages: number;
  onClick: () => void;
}

export interface ArrowProps {
  onClick: () => void;
  disabled?: boolean;
}

export interface PaginationInputProps {
  totalPages: number;
  onChange: (value: number) => void;
  onHandleKeyDown: (event: React.KeyboardEvent<HTMLInputElement>) => void;
  value?: number;
}

export const Pagination = ({ children, dataTestId }: BaseProps) => {
  return (
    <PaginationStyled data-testid={dataTestId ?? 'pagination-wrapper'}>
      {children}
    </PaginationStyled>
  );
};

Pagination.Content = ({ totalPages, onClick }: ContentProps) => {
  return (
    <Typography fontStyles='pXSmall'>
      of{' '}
      <LastPageStyled
        onClick={onClick}
        data-testid='pagination-last-page'
        title='Go to last page'
      >
        {totalPages}
      </LastPageStyled>
    </Typography>
  );
};

Pagination.NextItem = ({ onClick, disabled }: ArrowProps) => {
  return (
    <ArrowStyled
      title='Go to next page'
      onClick={onClick}
      disabled={disabled}
      data-testid='pagination-next-item'
    >
      <ArrowRightIcon />
    </ArrowStyled>
  );
};

Pagination.PreviousItem = ({ onClick, disabled }: ArrowProps) => {
  return (
    <ArrowStyled
      title='Go to previous page'
      onClick={onClick}
      disabled={disabled}
      data-testid='pagination-previous-item'
    >
      <ArrowLeftIcon />
    </ArrowStyled>
  );
};

export const PaginationInput = ({
  totalPages,
  onHandleKeyDown,
  value,
  onChange,
}: PaginationInputProps) => {
  const [inputValue, setInputValue] = useState(1);
  const [hasError, setHasError] = useState(false);

  const debouncedFn = useMemo(() => debounce((fn) => fn(), 500), []);

  useEffect(() => {
    if (value !== undefined) {
      setInputValue(value);
    }
  }, [value]);

  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const value = Number(event.target.value);
    setInputValue(value);

    const valueHasError = value > totalPages || value === 0;

    if (valueHasError) {
      setHasError(true);
    } else {
      setHasError(false);
    }

    debouncedFn(() => {
      if (valueHasError) return;
      onChange(value);
    });
  };

  return (
    <InputStyled
      data-testid='pagination-input'
      value={inputValue}
      onKeyDown={onHandleKeyDown}
      onChange={handleChange}
      hasError={hasError}
    />
  );
};
