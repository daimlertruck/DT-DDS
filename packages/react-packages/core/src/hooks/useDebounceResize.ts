import { useEffect } from 'react';

import { debounce } from '../utils';

export default function useDebounceResize(callback: () => void, delay = 500) {
  useEffect(() => {
    const debounceCallbackOnWindowResize = debounce(callback, delay);

    debounceCallbackOnWindowResize();
    window.addEventListener('resize', debounceCallbackOnWindowResize);
    return () =>
      window.removeEventListener('resize', debounceCallbackOnWindowResize);
  }, [callback, delay]);
}
