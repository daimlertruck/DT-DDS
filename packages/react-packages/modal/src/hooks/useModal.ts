import { useEffect, useState } from 'react';

const useModal = (initialState = false) => {
  const [isModalOpen, setIsModalOpen] = useState(!!initialState);

  useEffect(() => {
    setIsModalOpen(initialState);
  }, [initialState]);

  const toggleModal = () => {
    setIsModalOpen((prevState) => !prevState);
  };

  return {
    isModalOpen,
    toggleModal,
  };
};

export default useModal;
