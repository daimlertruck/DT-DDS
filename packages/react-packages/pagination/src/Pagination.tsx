import { BaseProps, debounce } from '@dt-dds/react-core';
import { Icon } from '@dt-dds/react-icon';
import { Typography } from '@dt-dds/react-typography';
import React, { useEffect, useMemo, useState } from 'react';

import {
  PaginationStyled,
  InputStyled,
  LastPageStyled,
  ArrowStyled,
} from './Pagination.styled';

export interface ContentProps {
  totalPages: number;
  onClick: () => void;
}

export interface ArrowProps {
  onClick: () => void;
  disabled?: boolean;
}

export interface PaginationInputProps {
  totalPages: number;
  onChange: (value: number) => void;
  onHandleKeyDown: (event: React.KeyboardEvent<HTMLInputElement>) => void;
  value?: number;
}

export const Pagination = ({ children, dataTestId }: BaseProps) => {
  return (
    <PaginationStyled data-testid={dataTestId ?? 'pagination-wrapper'}>
      {children}
    </PaginationStyled>
  );
};

Pagination.Content = ({ totalPages, onClick }: ContentProps) => {
  return (
    <Typography fontStyles='pXSmall'>
      of{' '}
      <LastPageStyled
        data-testid='pagination-last-page'
        onClick={onClick}
        title='Go to last page'
      >
        {totalPages}
      </LastPageStyled>
    </Typography>
  );
};

Pagination.NextItem = ({ onClick, disabled }: ArrowProps) => {
  return (
    <ArrowStyled
      data-testid='pagination-next-item'
      disabled={disabled}
      onClick={onClick}
      title='Go to next page'
    >
      <Icon code='arrow_forward_ios' size='s' />
    </ArrowStyled>
  );
};

Pagination.PreviousItem = ({ onClick, disabled }: ArrowProps) => {
  return (
    <ArrowStyled
      data-testid='pagination-previous-item'
      disabled={disabled}
      onClick={onClick}
      title='Go to previous page'
    >
      <Icon code='arrow_back_ios' size='s' />
    </ArrowStyled>
  );
};

export const PaginationInput = ({
  totalPages,
  onHandleKeyDown,
  value,
  onChange,
}: PaginationInputProps) => {
  const [inputValue, setInputValue] = useState(1);
  const [hasError, setHasError] = useState(false);

  const debouncedFn = useMemo(() => debounce((fn) => fn(), 500), []);

  useEffect(() => {
    if (value !== undefined) {
      setInputValue(value);
    }
  }, [value]);

  useEffect(() => {
    if (inputValue && inputValue <= totalPages) {
      debouncedFn(() => {
        onChange(inputValue);
      });
    }
  }, [debouncedFn, inputValue, onChange, totalPages]);

  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const value = Number(event.target.value);
    setInputValue(value);

    if (value > totalPages || value === 0) {
      setHasError(true);
    } else {
      setHasError(false);
    }
  };

  return (
    <InputStyled
      data-testid='pagination-input'
      hasError={hasError}
      onChange={handleChange}
      onKeyDown={onHandleKeyDown}
      value={inputValue}
    />
  );
};
