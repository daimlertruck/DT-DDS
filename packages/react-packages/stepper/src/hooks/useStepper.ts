import { useCallback, useState } from 'react';

export const useStepper = () => {
  const [activeStep, setActiveStep] = useState(0);

  const handleNext = () => {
    setActiveStep((activeStep) => activeStep + 1);
  };

  const handleBack = () => {
    setActiveStep((activeStep) => activeStep - 1);
  };

  const handleChangeStep = useCallback((step: number) => {
    setActiveStep(step);
  }, []);

  return {
    activeStep,
    handleChangeStep,
    handleNext,
    handleBack,
  };
};
