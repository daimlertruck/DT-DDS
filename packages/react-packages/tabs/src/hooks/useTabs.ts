import { useState, useEffect } from 'react';

export const useTabs = (initialTab: number = 0) => {
  const [activeTab, setActiveTab] = useState<number>(initialTab);
  useEffect(() => {
    setActiveTab(initialTab);
  }, [initialTab]);

  const handleChange = (value: number) => {
    setActiveTab(value);
  };

  return {
    activeTab,
    handleChange,
  };
};
