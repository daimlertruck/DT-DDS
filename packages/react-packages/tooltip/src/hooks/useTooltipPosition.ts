import { OTooltipDirection, TooltipDirection } from '../constants';
import { useTooltipContext } from '../context';

const PADDING = 4;
const ARROW_OFFSET = 8;
const OFFSET = PADDING + ARROW_OFFSET;
const DEFAULT_POSITION = { top: -1, left: -1 };

export const useTooltipPosition = (
  direction: TooltipDirection,
  tooltipRect: DOMRect
) => {
  const {
    containerRectProperties: { top, left, height, width },
  } = useTooltipContext();

  if (typeof window === 'undefined') return DEFAULT_POSITION;

  const scrollX = window.scrollX;
  const scrollY = window.scrollY;

  const positions = {
    [OTooltipDirection.Top]: (() => {
      const computedTop = top - OFFSET + scrollY;
      const computedLeft = left + width / 2 + scrollX;

      return {
        top: Math.max(computedTop, scrollY + tooltipRect.height),
        left: Math.min(
          scrollX + window.innerWidth - tooltipRect.width / 2,
          Math.max(scrollX + tooltipRect.width / 2, computedLeft)
        ),
      };
    })(),

    [OTooltipDirection.Bottom]: (() => {
      const computedTop = top + height + OFFSET + scrollY;
      const computedLeft = left + width / 2 + scrollX;

      return {
        top: Math.min(
          computedTop,
          scrollY + window.innerHeight - tooltipRect.height
        ),
        left: Math.min(
          scrollX + window.innerWidth - tooltipRect.width / 2,
          Math.max(scrollX + tooltipRect.width / 2, computedLeft)
        ),
      };
    })(),

    [OTooltipDirection.Left]: (() => {
      const computedTop = top + height / 2 + scrollY;
      const computedLeft = left - OFFSET + scrollX;

      return {
        top: Math.min(
          scrollY + window.innerHeight - tooltipRect.height / 2,
          Math.max(scrollY + tooltipRect.height / 2, computedTop)
        ),
        left: Math.max(computedLeft, scrollX + tooltipRect.width),
      };
    })(),

    [OTooltipDirection.Right]: (() => {
      const computedTop = top + height / 2 + scrollY;
      const computedLeft = left + width + OFFSET + scrollX;

      return {
        top: Math.min(
          scrollY + window.innerHeight - tooltipRect.height / 2,
          Math.max(scrollY + tooltipRect.height / 2, computedTop)
        ),
        left: Math.min(
          computedLeft,
          scrollX + window.innerWidth - tooltipRect.width
        ),
      };
    })(),
  };

  return positions[direction];
};
