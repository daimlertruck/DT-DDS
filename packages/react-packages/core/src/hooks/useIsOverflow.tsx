import { useTheme } from '@emotion/react';
import { useCallback, useRef, useState } from 'react';

import useDebounceResize from './useDebounceResize';
import useMedia from './useMedia';

export const useIsOverflow = <T extends HTMLElement>() => {
  const childrenRef = useRef<T>(null);
  const theme = useTheme();
  const isDesktop = useMedia(`(min-width: ${theme.breakpoints.mq3}px)`);
  const [overflow, setOverflow] = useState<boolean>(false);

  const truncateResize = useCallback(() => {
    if (!childrenRef.current || !isDesktop) {
      return;
    }

    const isOverflowed =
      childrenRef.current.scrollWidth > childrenRef.current.clientWidth;

    if (isOverflowed !== overflow) setOverflow(isOverflowed);
  }, [childrenRef, isDesktop, overflow]);

  useDebounceResize(truncateResize);

  return { overflow, childrenRef };
};
