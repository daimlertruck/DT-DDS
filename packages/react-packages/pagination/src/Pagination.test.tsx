import { withProviders } from '@dt-dds/react-core';
import {
  fireEvent,
  render,
  screen,
  waitFor,
  renderHook,
  act,
  RenderHookResult,
} from '@testing-library/react';
import React from 'react';

import { usePagination } from './hooks';
import { Pagination, PaginationInput } from './Pagination';

describe('<Pagination /> component', () => {
  const totalPages = 3;
  it('should render the pagination content', async () => {
    const currentPage = 1;
    const ProvidedPagination = withProviders(Pagination);
    const onClick = jest.fn();
    const { container } = render(
      <ProvidedPagination>
        <Pagination.PreviousItem disabled onClick={onClick} />
        <PaginationInput
          onChange={onClick}
          onHandleKeyDown={onClick}
          totalPages={totalPages}
          value={currentPage}
        />
        <Pagination.Content onClick={onClick} totalPages={totalPages} />
        <Pagination.NextItem onClick={onClick} />
      </ProvidedPagination>
    );

    expect(container).toMatchSnapshot();
  });

  describe('change page', () => {
    let hook: RenderHookResult<
      {
        currentPage: number;
        handleChange: (value: number) => void;
        handleUserInput: (e: React.KeyboardEvent<HTMLInputElement>) => void;
      },
      unknown
    >;
    beforeEach(() => {
      hook = renderHook(() => usePagination());
      const ProvidedPagination = withProviders(Pagination);
      const { result } = hook;

      render(
        <ProvidedPagination>
          <Pagination.PreviousItem
            onClick={() =>
              act(() => {
                result.current.handleChange(result.current.currentPage - 1);
              })
            }
          />
          <PaginationInput
            onChange={(value) => act(() => result.current.handleChange(value))}
            onHandleKeyDown={(e) =>
              act(() => {
                result.current.handleUserInput(e);
              })
            }
            totalPages={totalPages}
            value={result.current.currentPage}
          />
          <Pagination.Content
            onClick={() =>
              act(() => {
                result.current.handleChange(totalPages);
              })
            }
            totalPages={totalPages}
          />
          <Pagination.NextItem
            onClick={() =>
              act(() => {
                result.current.handleChange(result.current.currentPage + 1);
              })
            }
          />
        </ProvidedPagination>
      );
    });

    it('should increase page number when next item is clicked', () => {
      fireEvent.click(screen.getByTestId('pagination-next-item'));
      expect(hook.result.current.currentPage).toEqual(2);
    });

    it('should decrease page number when previous item is clicked', () => {
      act(() =>
        hook.result.current.handleChange(hook.result.current.currentPage + 1)
      );

      fireEvent.click(screen.getByTestId('pagination-previous-item'));
      expect(hook.result.current.currentPage).toEqual(1);
    });

    it('should change to last page when button is clicked', () => {
      fireEvent.click(screen.getByTestId('pagination-last-page'));
      expect(hook.result.current.currentPage).toEqual(3);
    });

    it('should change to page with number inserted on input', async () => {
      fireEvent.change(screen.getByTestId('pagination-input'), {
        target: { value: '2' },
      });
      await waitFor(() => expect(hook.result.current.currentPage).toEqual(2));
    });
  });
});
