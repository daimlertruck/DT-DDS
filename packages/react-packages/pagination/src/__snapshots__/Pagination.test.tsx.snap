// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination /> component should render the pagination content 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 8px;
  margin-top: 48px;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  background-color: #FFFFFF;
  border: none;
  border-radius: 3px;
  cursor: default;
}

.emotion-4 {
  font-family: DTUI-icons-outlined;
  font-weight: normal;
  font-style: normal;
  font-size: 16px;
  display: inline-block;
  line-height: 1;
  text-transform: none;
  letter-spacing: normal;
  word-wrap: normal;
  white-space: nowrap;
  font-variation-settings: 'FILL' 0;
  color: #292929;
  -webkit-font-smoothing: antialiased;
  text-rendering: optimizeLegibility;
  -moz-osx-font-smoothing: grayscale;
  font-feature-settings: 'liga';
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.emotion-4:hover,
.emotion-4:active {
  color: #00677F;
}

.emotion-6 {
  font-weight: 400;
  font-size: 14px;
  line-height: 18.76px;
  letter-spacing: normal;
  margin: 0;
  text-transform: none;
  text-align: center;
  padding: 6px 8px;
  margin: 0px;
  width: 50px;
  height: 32px;
  background-color: #FFFFFF;
  color: #292929;
  border: 1px solid #DEDEDE;
  border-radius: 3px;
}

.emotion-6:focus {
  border-bottom: 2px solid #000000;
}

.emotion-8 {
  font-weight: 400;
  font-size: 14px;
  line-height: 18.76px;
  letter-spacing: normal;
  margin: 0;
  text-transform: none;
  color: #292929;
}

.emotion-10 {
  background-color: transparent;
  color: #000000;
  border: none;
  cursor: pointer;
}

.emotion-12 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  background-color: #FFFFFF;
  border: none;
  border-radius: 3px;
  cursor: pointer;
}

<div>
  <div
    class="emotion-0 emotion-1"
    data-testid="pagination-wrapper"
  >
    <button
      class="emotion-2 emotion-3"
      data-testid="pagination-previous-item"
      disabled=""
      title="Go to previous page"
    >
      <i
        class="emotion-4 emotion-5"
        color="#292929"
        data-testid="icon"
      >
        arrow_back_ios
      </i>
    </button>
    <input
      class="emotion-6 emotion-7"
      data-testid="pagination-input"
      value="1"
    />
    <p
      class="emotion-8 emotion-9"
      color="grey_90"
      data-testid="typography"
    >
      of
       
      <button
        class="emotion-10 emotion-11"
        data-testid="pagination-last-page"
        title="Go to last page"
      >
        3
      </button>
    </p>
    <button
      class="emotion-12 emotion-3"
      data-testid="pagination-next-item"
      title="Go to next page"
    >
      <i
        class="emotion-4 emotion-5"
        color="#292929"
        data-testid="icon"
      >
        arrow_forward_ios
      </i>
    </button>
  </div>
</div>
`;
